File: hydra-python/src/main/python/hydra/lib/flows.py
"""Python implementations of hydra.lib.flows primitives."""

from collections.abc import Callable, Mapping, Sequence
from typing import Any

from hydra.compute import Flow, FlowState, Trace
from hydra.dsl.python import FrozenDict, frozenlist, Maybe, Just, Nothing, NOTHING


def apply[S, X, Y](f: Flow[S, Callable[[X], Y]], x: Flow[S, X]) -> Flow[S, Y]:
    """Apply a function in a Flow to a value in a Flow (applicative)."""
    def run(state: S, trace: Trace) -> FlowState[S, Y]:
        f_state = f.value(state, trace)

        match f_state.value:
            case Nothing():
                return FlowState(value=NOTHING, state=f_state.state, trace=f_state.trace)
            case Just(func):
                x_state = x.value(f_state.state, f_state.trace)

                match x_state.value:
                    case Nothing():
                        return FlowState(value=NOTHING, state=x_state.state, trace=x_state.trace)
                    case Just(val):
                        result = func(val)
                        return FlowState(value=Just(result), state=x_state.state, trace=x_state.trace)

    return Flow(run)


def bind[S, X, Y](mx: Flow[S, X], f: Callable[[X], Flow[S, Y]]) -> Flow[S, Y]:
    """Monadic bind for Flow."""
    def run(state: S, trace: Trace) -> FlowState[S, Y]:
        x_state = mx.value(state, trace)

        match x_state.value:
            case Nothing():
                return FlowState(value=NOTHING, state=x_state.state, trace=x_state.trace)
            case Just(x):
                my = f(x)
                return my.value(x_state.state, x_state.trace)

    return Flow(run)


def fail(message: str) -> Flow[Any, Any]:
    """Create a failed Flow with an error message."""
    def run(state: Any, trace: Trace) -> FlowState[Any, Any]:
        new_messages = tuple(list(trace.messages) + [message])
        new_trace = Trace(
            stack=trace.stack,
            messages=new_messages,
            other=trace.other
        )
        return FlowState(value=NOTHING, state=state, trace=new_trace)

    return Flow(run)


def foldl[S, A, B](f: Callable[[A, B], Flow[S, A]], initial: A, values: Sequence[B]) -> Flow[S, A]:
    """Fold over a list with a monadic function."""
    def run(state: S, trace: Trace) -> FlowState[S, A]:
        current_state = state
        current_trace = trace
        current_value = initial

        for b in values:
            flow_result = f(current_value, b)
            result_state = flow_result.value(current_state, current_trace)

            match result_state.value:
                case Nothing():
                    return FlowState(value=NOTHING, state=result_state.state, trace=result_state.trace)
                case Just(val):
                    current_value = val
                    current_state = result_state.state
                    current_trace = result_state.trace

        return FlowState(value=Just(current_value), state=current_state, trace=current_trace)

    return Flow(run)


def map[S, X, Y](f: Callable[[X], Y], mx: Flow[S, X]) -> Flow[S, Y]:
    """Map a function over a Flow (functor)."""
    def run(state: S, trace: Trace) -> FlowState[S, Y]:
        x_state = mx.value(state, trace)

        match x_state.value:
            case Nothing():
                return FlowState(value=NOTHING, state=x_state.state, trace=x_state.trace)
            case Just(x):
                result = f(x)
                return FlowState(value=Just(result), state=x_state.state, trace=x_state.trace)

    return Flow(run)


def map_elems[S, K, V1, V2](f: Callable[[V1], Flow[S, V2]], m: Mapping[K, V1]) -> Flow[S, FrozenDict[K, V2]]:
    """Map a monadic function over the values of a map."""
    def run(state: S, trace: Trace) -> FlowState[S, FrozenDict[K, V2]]:
        current_state = state
        current_trace = trace
        result_dict: dict[K, V2] = {}

        for k, v1 in m.items():
            flow_result = f(v1)
            v2_state = flow_result.value(current_state, current_trace)

            match v2_state.value:
                case Nothing():
                    return FlowState(value=NOTHING, state=v2_state.state, trace=v2_state.trace)
                case Just(v2):
                    result_dict[k] = v2
                    current_state = v2_state.state
                    current_trace = v2_state.trace

        return FlowState(value=Just(FrozenDict(result_dict)), state=current_state, trace=current_trace)

    return Flow(run)


def map_keys[S, K1, K2, V](f: Callable[[K1], Flow[S, K2]], m: Mapping[K1, V]) -> Flow[S, FrozenDict[K2, V]]:
    """Map a monadic function over the keys of a map."""
    def run(state: S, trace: Trace) -> FlowState[S, FrozenDict[K2, V]]:
        current_state = state
        current_trace = trace
        result_dict: dict[K2, V] = {}

        for k1, v in m.items():
            flow_result = f(k1)
            k2_state = flow_result.value(current_state, current_trace)

            match k2_state.value:
                case Nothing():
                    return FlowState(value=NOTHING, state=k2_state.state, trace=k2_state.trace)
                case Just(k2):
                    result_dict[k2] = v
                    current_state = k2_state.state
                    current_trace = k2_state.trace

        return FlowState(value=Just(FrozenDict(result_dict)), state=current_state, trace=current_trace)

    return Flow(run)


def map_list[S, X, Y](f: Callable[[X], Flow[S, Y]], xs: Sequence[X]) -> Flow[S, frozenlist[Y]]:
    """Map a monadic function over a list."""
    def run(state: S, trace: Trace) -> FlowState[S, frozenlist[Y]]:
        current_state = state
        current_trace = trace
        results: list[Y] = []

        for x in xs:
            flow_result = f(x)
            y_state = flow_result.value(current_state, current_trace)

            match y_state.value:
                case Nothing():
                    return FlowState(value=NOTHING, state=y_state.state, trace=y_state.trace)
                case Just(y):
                    results.append(y)
                    current_state = y_state.state
                    current_trace = y_state.trace

        return FlowState(value=Just(tuple(results)), state=current_state, trace=current_trace)

    return Flow(run)


def map_maybe[S, X, Y](f: Callable[[X], Flow[S, Y]], mx: Maybe[X]) -> Flow[S, Maybe[Y]]:
    """Map a monadic function over an optional value."""
    def run(state: S, trace: Trace) -> FlowState[S, Maybe[Y]]:
        match mx:
            case Nothing():
                # Input is Nothing, succeed with Nothing as result
                # This properly represents "Just Nothing" - a successful computation with no value
                return FlowState(value=Just(NOTHING), state=state, trace=trace)
            case Just(x):
                # Input is Just x, apply the function
                flow_result = f(x)
                y_state = flow_result.value(state, trace)

                match y_state.value:
                    case Nothing():
                        # Flow failed - return Nothing (not Just(Nothing))
                        return FlowState(value=NOTHING, state=y_state.state, trace=y_state.trace)
                    case Just(y):
                        # Flow succeeded with value - wrap in Just(Just(...))
                        return FlowState(value=Just(Just(y)), state=y_state.state, trace=y_state.trace)

    return Flow(run)


def map_set[S, X, Y](f: Callable[[X], Flow[S, Y]], xs: frozenset[X]) -> Flow[S, frozenset[Y]]:
    """Map a monadic function over a set."""
    def run(state: S, trace: Trace) -> FlowState[S, frozenset[Y]]:
        current_state = state
        current_trace = trace
        results: list[Y] = []

        for x in xs:
            flow_result = f(x)
            y_state = flow_result.value(current_state, current_trace)

            match y_state.value:
                case Nothing():
                    return FlowState(value=NOTHING, state=y_state.state, trace=y_state.trace)
                case Just(y):
                    results.append(y)
                    current_state = y_state.state
                    current_trace = y_state.trace

        return FlowState(value=Just(frozenset(results)), state=current_state, trace=current_trace)

    return Flow(run)


def pure[X](x: X) -> Flow[Any, X]:
    """Lift a value into a Flow."""
    def run(state: Any, trace: Trace) -> FlowState[Any, X]:
        return FlowState(value=Just(x), state=state, trace=trace)

    return Flow(run)


def sequence[S, X](flows: Sequence[Flow[S, X]]) -> Flow[S, frozenlist[X]]:
    """Sequence a list of Flows into a Flow of a list."""
    def run(state: S, trace: Trace) -> FlowState[S, frozenlist[X]]:
        current_state = state
        current_trace = trace
        results: list[X] = []

        for flow in flows:
            x_state = flow.value(current_state, current_trace)

            match x_state.value:
                case Nothing():
                    return FlowState(value=NOTHING, state=x_state.state, trace=x_state.trace)
                case Just(x):
                    results.append(x)
                    current_state = x_state.state
                    current_trace = x_state.trace

        return FlowState(value=Just(tuple(results)), state=current_state, trace=current_trace)

    return Flow(run)

File: hydra-haskell/src/main/haskell/Hydra/Sources/Eval/Lib/Flows.hs

module Hydra.Sources.Eval.Lib.Flows where

-- Standard imports for kernel terms modules
import Hydra.Kernel
import Hydra.Sources.Libraries
import qualified Hydra.Dsl.Meta.Accessors     as Accessors
import qualified Hydra.Dsl.Annotations   as Annotations
import qualified Hydra.Dsl.Meta.Ast           as Ast
import qualified Hydra.Dsl.Bootstrap     as Bootstrap
import qualified Hydra.Dsl.Meta.Coders        as Coders
import qualified Hydra.Dsl.Meta.Compute       as Compute
import qualified Hydra.Dsl.Meta.Core          as Core
import qualified Hydra.Dsl.Meta.Grammar       as Grammar
import qualified Hydra.Dsl.Grammars      as Grammars
import qualified Hydra.Dsl.Meta.Graph         as Graph
import qualified Hydra.Dsl.Meta.Json          as Json
import qualified Hydra.Dsl.Meta.Lib.Chars     as Chars
import qualified Hydra.Dsl.Meta.Lib.Eithers   as Eithers
import qualified Hydra.Dsl.Meta.Lib.Equality  as Equality
import qualified Hydra.Dsl.Meta.Lib.Flows     as Flows
import qualified Hydra.Dsl.Meta.Lib.Lists     as Lists
import qualified Hydra.Dsl.Meta.Lib.Literals  as Literals
import qualified Hydra.Dsl.Meta.Lib.Logic     as Logic
import qualified Hydra.Dsl.Meta.Lib.Maps      as Maps
import qualified Hydra.Dsl.Meta.Lib.Math      as Math
import qualified Hydra.Dsl.Meta.Lib.Maybes    as Maybes
import qualified Hydra.Dsl.Meta.Lib.Pairs     as Pairs
import qualified Hydra.Dsl.Meta.Lib.Sets      as Sets
import           Hydra.Dsl.Meta.Lib.Strings   as Strings
import qualified Hydra.Dsl.Literals      as Literals
import qualified Hydra.Dsl.LiteralTypes  as LiteralTypes
import qualified Hydra.Dsl.Meta.Base     as MetaBase
import qualified Hydra.Dsl.Meta.Terms    as MetaTerms
import qualified Hydra.Dsl.Meta.Types    as MetaTypes
import qualified Hydra.Dsl.Meta.Module        as Module
import           Hydra.Dsl.Meta.Phantoms as Phantoms
import qualified Hydra.Dsl.Prims         as Prims
import qualified Hydra.Dsl.Tabular       as Tabular
import qualified Hydra.Dsl.Meta.Testing       as Testing
import qualified Hydra.Dsl.Terms         as Terms
import qualified Hydra.Dsl.Meta.Testing       as Testing
import qualified Hydra.Dsl.Tests         as Tests
import qualified Hydra.Dsl.Meta.Topology      as Topology
import qualified Hydra.Dsl.Types         as Types
import qualified Hydra.Dsl.Meta.Typing        as Typing
import qualified Hydra.Dsl.Meta.Util          as Util
import qualified Hydra.Dsl.Meta.Variants      as Variants
import           Hydra.Sources.Kernel.Types.All
import           Prelude hiding ((++), map)
import qualified Data.Int                as I
import qualified Data.List               as L
import qualified Data.Map                as M
import qualified Data.Set                as S
import qualified Data.Maybe              as Y

import qualified Hydra.Sources.Kernel.Terms.Extract.Core as ExtractCore
import qualified Hydra.Sources.Kernel.Terms.Monads as Monads
import qualified Hydra.Sources.Kernel.Terms.Show.Core as ShowCore


ns :: Namespace
ns = Namespace "hydra.eval.lib.flows"

define :: String -> TTerm a -> TBinding a
define = definitionInNamespace ns

module_ :: Module
module_ = Module ns elements
    [ExtractCore.ns, Monads.ns, ShowCore.ns]
    kernelTypesNamespaces $
    Just ("Evaluation-level implementations of Flow functions for the Hydra interpreter.")
  where
    elements = [
      toBinding apply_,
      toBinding bind_,
      toBinding foldl_,
      toBinding map_,
      toBinding mapElems_,
      toBinding mapKeys_,
      toBinding mapList_,
      toBinding mapMaybe_,
      toBinding mapSet_]

-- | Interpreter-friendly applicative apply for Flow.
-- apply flowFun flowArg: applies the function inside flowFun to the value inside flowArg.
apply_ :: TBinding (Term -> Term -> Flow s Term)
apply_ = define "apply" $
  doc "Interpreter-friendly applicative apply for Flow." $
  "flowFun" ~> "flowArg" ~>
  -- Build: flowFun >>= \f -> flowArg >>= \x -> pure (f x)
  produce $ Core.termApplication $ Core.application
    (Core.termApplication $ Core.application
      (Core.termFunction $ Core.functionPrimitive $ encodedName _flows_bind)
      (var "flowFun"))
    (Core.termFunction $ Core.functionLambda $ Core.lambda (wrap _Name $ string "f") nothing $
      Core.termApplication $ Core.application
        (Core.termApplication $ Core.application
          (Core.termFunction $ Core.functionPrimitive $ encodedName _flows_bind)
          (var "flowArg"))
        (Core.termFunction $ Core.functionLambda $ Core.lambda (wrap _Name $ string "x") nothing $
          Core.termApplication $ Core.application
            (Core.termFunction $ Core.functionPrimitive $ encodedName _flows_pure)
            (Core.termApplication $ Core.application
              (Core.termVariable $ wrap _Name $ string "f")
              (Core.termVariable $ wrap _Name $ string "x"))))

-- | Interpreter-friendly monadic bind for Flow.
-- Pattern matches on the flow term to extract the wrapped function,
-- then builds a new Flow that sequences the computation.
bind_ :: TBinding (Term -> Term -> Flow s Term)
bind_ = define "bind" $
  doc "Interpreter-friendly monadic bind for Flow." $
  "flowTerm" ~> "funTerm" ~>
  -- Pattern match on flowTerm to see if it's a TermWrap (Flow)
  cases _Term (var "flowTerm")
    (Just (Monads.unexpected @@ string "flow term" @@ (ShowCore.term @@ var "flowTerm"))) [
    _Term_wrap>>: "wrappedTerm" ~>
      "innerFun" <~ Core.wrappedTermBody (var "wrappedTerm") $
      produce $ Core.termWrap $ Core.wrappedTerm
        (wrap _Name $ string "hydra.compute.Flow")
        (Core.termFunction $ Core.functionLambda $ Core.lambda (wrap _Name $ string "s") nothing $
          Core.termFunction $ Core.functionLambda $ Core.lambda (wrap _Name $ string "t") nothing $
            -- Build: let fs = innerFun s t
            --        in maybe (FlowState Nothing fs.state fs.trace)
            --                 (\v -> unwrap (funTerm v) fs.state fs.trace)
            --                 fs.value
            Core.termApplication $ Core.application
              (Core.termApplication $ Core.application
                (Core.termApplication $ Core.application
                  (Core.termFunction $ Core.functionPrimitive $ encodedName _maybes_maybe)
                  -- default (Nothing): return FlowState with Nothing
                  (Core.termRecord $ Core.record (wrap _Name $ string "hydra.compute.FlowState") $
                    list [
                      Core.field (wrap _Name $ string "value") (Core.termMaybe nothing),
                      Core.field (wrap _Name $ string "state") (projectFs (string "state")),
                      Core.field (wrap _Name $ string "trace") (projectFs (string "trace"))]))
                -- Just case: \v -> unwrap (funTerm v) fs.state fs.trace
                (Core.termFunction $ Core.functionLambda $ Core.lambda (wrap _Name $ string "v") nothing $
                  Core.termApplication $ Core.application
                    (Core.termApplication $ Core.application
                      (Core.termFunction $ Core.functionElimination $ Core.eliminationWrap $ wrap _Name $ string "hydra.compute.Flow")
                      (Core.termApplication $ Core.application (var "funTerm") (Core.termVariable $ wrap _Name $ string "v")))
                    (Core.termApplication $ Core.application
                      (projectFs (string "state"))
                      (projectFs (string "trace")))))
              -- fs.value
              (projectFs (string "value")))]
  where
    -- Run the inner function to get FlowState: innerFun s t
    runFlow :: TTerm Term
    runFlow = Core.termApplication $ Core.application
      (Core.termApplication $ Core.application
        (var "innerFun")
        (Core.termVariable $ wrap _Name $ string "s"))
      (Core.termVariable $ wrap _Name $ string "t")
    -- Project a field from the FlowState
    projectFs :: TTerm String -> TTerm Term
    projectFs fieldName = Core.termApplication $ Core.application
      (Core.termFunction $ Core.functionElimination $ Core.eliminationRecord $ Core.projection
        (wrap _Name $ string "hydra.compute.FlowState")
        (wrap _Name fieldName))
      runFlow

-- | Interpreter-friendly foldl for Flow.
-- foldl funTerm initTerm listTerm: folds over the list with Flow effects.
foldl_ :: TBinding (Term -> Term -> Term -> Flow s Term)
foldl_ = define "foldl" $
  doc "Interpreter-friendly foldl for Flow." $
  "funTerm" ~> "initTerm" ~> "listTerm" ~>
  "elements" <<~ ExtractCore.list @@ var "listTerm" $
  -- Build: fold over elements using bind
  -- foldl f init [e1,e2,...] = bind (f init e1) (\acc1 -> bind (f acc1 e2) (\acc2 -> ...))
  produce $ Lists.foldl
    ("acc" ~> "el" ~>
      Core.termApplication $ Core.application
        (Core.termApplication $ Core.application
          (Core.termFunction $ Core.functionPrimitive $ encodedName _flows_bind)
          (var "acc"))
        (Core.termFunction $ Core.functionLambda $ Core.lambda (wrap _Name $ string "accVal") nothing $
          Core.termApplication $ Core.application
            (Core.termApplication $ Core.application (var "funTerm") (Core.termVariable $ wrap _Name $ string "accVal"))
            (var "el")))
    (Core.termApplication $ Core.application
      (Core.termFunction $ Core.functionPrimitive $ encodedName _flows_pure)
      (var "initTerm"))
    (var "elements")

-- | Interpreter-friendly functor map for Flow.
-- map funTerm flowTerm: applies funTerm to the value inside flowTerm.
map_ :: TBinding (Term -> Term -> Flow s Term)
map_ = define "map" $
  doc "Interpreter-friendly functor map for Flow." $
  "funTerm" ~> "flowTerm" ~>
  -- Pattern match on flowTerm to see if it's a TermWrap (Flow)
  cases _Term (var "flowTerm")
    (Just (Monads.unexpected @@ string "flow term" @@ (ShowCore.term @@ var "flowTerm"))) [
    _Term_wrap>>: "wrappedTerm" ~>
      "innerFun" <~ Core.wrappedTermBody (var "wrappedTerm") $
      produce $ Core.termWrap $ Core.wrappedTerm
        (wrap _Name $ string "hydra.compute.Flow")
        (Core.termFunction $ Core.functionLambda $ Core.lambda (wrap _Name $ string "s") nothing $
          Core.termFunction $ Core.functionLambda $ Core.lambda (wrap _Name $ string "t") nothing $
            -- Build: let fs = innerFun s t
            --        in FlowState (map (\v -> funTerm v) fs.value) fs.state fs.trace
            Core.termRecord $ Core.record (wrap _Name $ string "hydra.compute.FlowState") $
              list [
                Core.field (wrap _Name $ string "value")
                  (Core.termApplication $ Core.application
                    (Core.termApplication $ Core.application
                      (Core.termFunction $ Core.functionPrimitive $ encodedName _maybes_map)
                      (Core.termFunction $ Core.functionLambda $ Core.lambda (wrap _Name $ string "v") nothing $
                        Core.termApplication $ Core.application (var "funTerm") (Core.termVariable $ wrap _Name $ string "v")))
                    (projectFsMap (string "value"))),
                Core.field (wrap _Name $ string "state") (projectFsMap (string "state")),
                Core.field (wrap _Name $ string "trace") (projectFsMap (string "trace"))])]
  where
    -- Run the inner function to get FlowState: innerFun s t
    runFlowMap :: TTerm Term
    runFlowMap = Core.termApplication $ Core.application
      (Core.termApplication $ Core.application
        (var "innerFun")
        (Core.termVariable $ wrap _Name $ string "s"))
      (Core.termVariable $ wrap _Name $ string "t")
    -- Project a field from the FlowState
    projectFsMap :: TTerm String -> TTerm Term
    projectFsMap fieldName = Core.termApplication $ Core.application
      (Core.termFunction $ Core.functionElimination $ Core.eliminationRecord $ Core.projection
        (wrap _Name $ string "hydra.compute.FlowState")
        (wrap _Name fieldName))
      runFlowMap

-- | Interpreter-friendly mapElems for Map with Flow.
-- mapElems funTerm mapTerm: applies funTerm to each value in the map.
mapElems_ :: TBinding (Term -> Term -> Flow s Term)
mapElems_ = define "mapElems" $
  doc "Interpreter-friendly mapElems for Map with Flow." $
  "funTerm" ~> "mapTerm" ~>
  cases _Term (var "mapTerm")
    (Just (Monads.unexpected @@ string "map value" @@ (ShowCore.term @@ var "mapTerm"))) [
    _Term_map>>: "m" ~>
      "pairs" <~ Maps.toList (var "m") $
      -- Build: sequence (map (\(k,v) -> map (\v' -> (k, v')) (funTerm v)) pairs) >>= fromList
      produce $ Core.termApplication $ Core.application
        (Core.termApplication $ Core.application
          (Core.termFunction $ Core.functionPrimitive $ encodedName _flows_bind)
          (Core.termApplication $ Core.application
            (Core.termFunction $ Core.functionPrimitive $ encodedName _flows_sequence)
            (Core.termList $ Lists.map
              ("p" ~>
                "k" <~ Pairs.first (var "p") $
                "v" <~ Pairs.second (var "p") $
                Core.termApplication $ Core.application
                  (Core.termApplication $ Core.application
                    (Core.termFunction $ Core.functionPrimitive $ encodedName _flows_map)
                    (Core.termFunction $ Core.functionLambda $ Core.lambda (wrap _Name $ string "v'") nothing $
                      Core.termPair $ pair (var "k") (Core.termVariable $ wrap _Name $ string "v'")))
                  (Core.termApplication $ Core.application (var "funTerm") (var "v")))
              (var "pairs"))))
        (Core.termFunction $ Core.functionLambda $ Core.lambda (wrap _Name $ string "newPairs") nothing $
          Core.termApplication $ Core.application
            (Core.termFunction $ Core.functionPrimitive $ encodedName _flows_pure)
            (Core.termApplication $ Core.application
              (Core.termFunction $ Core.functionPrimitive $ encodedName _maps_fromList)
              (Core.termVariable $ wrap _Name $ string "newPairs")))]

-- | Interpreter-friendly mapKeys for Map with Flow.
-- mapKeys funTerm mapTerm: applies funTerm to each key in the map.
mapKeys_ :: TBinding (Term -> Term -> Flow s Term)
mapKeys_ = define "mapKeys" $
  doc "Interpreter-friendly mapKeys for Map with Flow." $
  "funTerm" ~> "mapTerm" ~>
  cases _Term (var "mapTerm")
    (Just (Monads.unexpected @@ string "map value" @@ (ShowCore.term @@ var "mapTerm"))) [
    _Term_map>>: "m" ~>
      "pairs" <~ Maps.toList (var "m") $
      -- Build: sequence (map (\(k,v) -> map (\k' -> (k', v)) (funTerm k)) pairs) >>= fromList
      produce $ Core.termApplication $ Core.application
        (Core.termApplication $ Core.application
          (Core.termFunction $ Core.functionPrimitive $ encodedName _flows_bind)
          (Core.termApplication $ Core.application
            (Core.termFunction $ Core.functionPrimitive $ encodedName _flows_sequence)
            (Core.termList $ Lists.map
              ("p" ~>
                "k" <~ Pairs.first (var "p") $
                "v" <~ Pairs.second (var "p") $
                Core.termApplication $ Core.application
                  (Core.termApplication $ Core.application
                    (Core.termFunction $ Core.functionPrimitive $ encodedName _flows_map)
                    (Core.termFunction $ Core.functionLambda $ Core.lambda (wrap _Name $ string "k'") nothing $
                      Core.termPair $ pair (Core.termVariable $ wrap _Name $ string "k'") (var "v")))
                  (Core.termApplication $ Core.application (var "funTerm") (var "k")))
              (var "pairs"))))
        (Core.termFunction $ Core.functionLambda $ Core.lambda (wrap _Name $ string "newPairs") nothing $
          Core.termApplication $ Core.application
            (Core.termFunction $ Core.functionPrimitive $ encodedName _flows_pure)
            (Core.termApplication $ Core.application
              (Core.termFunction $ Core.functionPrimitive $ encodedName _maps_fromList)
              (Core.termVariable $ wrap _Name $ string "newPairs")))]

-- | Interpreter-friendly mapList for List with Flow (traverse).
-- mapList funTerm listTerm: applies funTerm to each element, collecting results.
mapList_ :: TBinding (Term -> Term -> Flow s Term)
mapList_ = define "mapList" $
  doc "Interpreter-friendly mapList for List with Flow." $
  "funTerm" ~> "listTerm" ~>
  "elements" <<~ ExtractCore.list @@ var "listTerm" $
  -- Build: sequence (map funTerm elements)
  produce $ Core.termApplication $ Core.application
    (Core.termFunction $ Core.functionPrimitive $ encodedName _flows_sequence)
    (Core.termList $ Lists.map
      ("el" ~> Core.termApplication $ Core.application (var "funTerm") (var "el"))
      (var "elements"))

-- | Interpreter-friendly mapMaybe for Maybe with Flow (traverse).
-- mapMaybe funTerm maybeTerm: applies funTerm if Just, returns Nothing otherwise.
mapMaybe_ :: TBinding (Term -> Term -> Flow s Term)
mapMaybe_ = define "mapMaybe" $
  doc "Interpreter-friendly mapMaybe for Maybe with Flow." $
  "funTerm" ~> "maybeTerm" ~>
  cases _Term (var "maybeTerm")
    (Just (Monads.unexpected @@ string "optional value" @@ (ShowCore.term @@ var "maybeTerm"))) [
    _Term_maybe>>: "m" ~>
      -- Use ifElse with term-level lambdas to delay evaluation (same pattern as Maybes)
      produce $ Core.termApplication $ Core.application
        (Core.termApplication $ Core.application
          (Core.termApplication $ Core.application
            (Core.termApplication $ Core.application
              (Core.termFunction $ Core.functionPrimitive $ encodedName _logic_ifElse)
              -- isNothing m
              (Core.termApplication $ Core.application
                (Core.termFunction $ Core.functionPrimitive $ encodedName _maybes_isNothing)
                (Core.termMaybe $ var "m")))
            -- Lambda returning pure Nothing (delayed)
            (Core.termFunction $ Core.functionLambda $ Core.lambda (wrap _Name $ string "_") nothing $
              Core.termApplication $ Core.application
                (Core.termFunction $ Core.functionPrimitive $ encodedName _flows_pure)
                (Core.termMaybe nothing)))
          -- Lambda returning map Just (funTerm (fromJust m)) (delayed)
          (Core.termFunction $ Core.functionLambda $ Core.lambda (wrap _Name $ string "_") nothing $
            Core.termApplication $ Core.application
              (Core.termApplication $ Core.application
                (Core.termFunction $ Core.functionPrimitive $ encodedName _flows_map)
                (Core.termFunction $ Core.functionLambda $ Core.lambda (wrap _Name $ string "x") nothing $
                  Core.termMaybe $ just $ Core.termVariable $ wrap _Name $ string "x"))
              (Core.termApplication $ Core.application
                (var "funTerm")
                (Core.termApplication $ Core.application
                  (Core.termFunction $ Core.functionPrimitive $ encodedName _maybes_fromJust)
                  (Core.termMaybe $ var "m")))))
        Core.termUnit]

-- | Interpreter-friendly mapSet for Set with Flow (traverse).
-- mapSet funTerm setTerm: applies funTerm to each element, collecting results.
mapSet_ :: TBinding (Term -> Term -> Flow s Term)
mapSet_ = define "mapSet" $
  doc "Interpreter-friendly mapSet for Set with Flow." $
  "funTerm" ~> "setTerm" ~>
  "elements" <<~ ExtractCore.set @@ var "setTerm" $
  -- Build: map fromList (sequence (map funTerm (toList elements)))
  produce $ Core.termApplication $ Core.application
    (Core.termApplication $ Core.application
      (Core.termFunction $ Core.functionPrimitive $ encodedName _flows_map)
      (Core.termFunction $ Core.functionPrimitive $ encodedName _sets_fromList))
    (Core.termApplication $ Core.application
      (Core.termFunction $ Core.functionPrimitive $ encodedName _flows_sequence)
      (Core.termList $ Lists.map
        ("el" ~> Core.termApplication $ Core.application (var "funTerm") (var "el"))
        (Sets.toList $ var "elements")))
